components/raftstore-v2/src/operation/life.rs

```diff
@@ -107,6 +110,66 @@ impl DestroyProgress {
     }
 }
 
+#[derive(Default)]
+pub struct AbnormalPeerContext {
+    /// Record the instants of peers being added into the configuration.
+    /// Remove them after they are not pending any more.
+    peers_start_pending_time: Vec<(u64, Instant)>,
+    /// A inaccurate cache about which peer is marked as down.
+    down_peers: Vec<u64>,
+}
+
+impl AbnormalPeerContext {
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.peers_start_pending_time.is_empty() && self.down_peers.is_empty()
+    }
+
+    #[inline]
+    pub fn reset(&mut self) {
+        self.peers_start_pending_time.clear();
+        self.down_peers.clear();
+    }
+
+    #[inline]
+    pub fn is_peer_down(&self, peer_id: u64) -> bool {
+        self.down_peers.contains(&peer_id)
+    }
+
+    #[inline]
+    pub fn is_peer_pending(&self, peer_id: u64) -> bool {
+        self.peers_start_pending_time.iter().any(|p| p.0 == peer_id)
+    }
+
+    #[inline]
+    pub fn collect_down_peers(&mut self, peer_ids: Vec<u64>) {
+        self.down_peers = peer_ids;
+    }
+
+    #[inline]
+    pub fn append_pending_peers(&mut self, mut peers: Vec<(u64, Instant)>) {
+        self.peers_start_pending_time.append(&mut peers);
+    }
```

Comment:
I think you could remove these super short functions that are used only once. (make them pub)

Reply:
I guess is as same as we use inline here?

Reply:
It's not a performance concern. New function requires reading and understanding, but directly accessing the std container is easy to understand.

Reply:
Ok, I can change it, BTW @BusyJay WDYT?

Reply:
You can expose necessary accessors.